File: dart/builder.dart
// Product: Pizza
class Pizza {
  String crust;
  String sauce;
  String cheese;
  List<String> toppings;

  Pizza() {
    crust = '';
    sauce = '';
    cheese = '';
    toppings = [];
  }

  void describe() {
    print('Pizza: $crust, $sauce, $cheese, Toppings: ${toppings.join(', ')}');
  }
}

// Builder Interface
abstract class PizzaBuilder {
  void buildCrust();
  void buildSauce();
  void buildCheese();
  void buildToppings();
  Pizza getPizza();
}

// Concrete Builder: MargheritaPizzaBuilder
class MargheritaPizzaBuilder implements PizzaBuilder {
  Pizza _pizza = Pizza();

  @override
  void buildCrust() {
    _pizza.crust = 'Thin Crust';
  }

  @override
  void buildSauce() {
    _pizza.sauce = 'Tomato Sauce';
  }

  @override
  void buildCheese() {
    _pizza.cheese = 'Mozzarella Cheese';
  }

  @override
  void buildToppings() {
    _pizza.toppings = ['Tomatoes', 'Basil'];
  }

  @override
  Pizza getPizza() {
    return _pizza;
  }
}

// Director: PizzaDirector
class PizzaDirector {
  PizzaBuilder builder;

  PizzaDirector(this.builder);

  void construct() {
    builder.buildCrust();
    builder.buildSauce();
    builder.buildCheese();
    builder.buildToppings();
  }
}

// Client
void main() {
  var margheritaBuilder = MargheritaPizzaBuilder();
  var pizzaDirector = PizzaDirector(margheritaBuilder);
  
  pizzaDirector.construct();
  var margheritaPizza = margheritaBuilder.getPizza();
  
  margheritaPizza.describe();
}

File: javascript/builder.js
// Product
class Pizza {
    constructor() {
        this.size = "";
        this.cheese = false;
        this.pepperoni = false;
        this.mushrooms = false;
    }

    toString() {
        return `Pizza: Size(${this.size}), Cheese(${this.cheese}), Pepperoni(${this.pepperoni}), Mushrooms(${this.mushrooms})`;
    }
}

// Builder Interface
class PizzaBuilder {
    constructor() {
        this.pizza = new Pizza();
    }

    buildSize() {}

    buildCheese() {}

    buildPepperoni() {}

    buildMushrooms() {}

    getResult() {
        return this.pizza;
    }
}

// Concrete Builder
class MargheritaPizzaBuilder extends PizzaBuilder {
    buildSize() {
        this.pizza.size = "Medium";
    }

    buildCheese() {
        this.pizza.cheese = true;
    }

    buildPepperoni() {
        // Margherita pizza doesn't have pepperoni
    }

    buildMushrooms() {
        // Margherita pizza doesn't have mushrooms
    }
}

// Director
class PizzaDirector {
    constructor(pizzaBuilder) {
        this.pizzaBuilder = pizzaBuilder;
    }

    constructPizza() {
        this.pizzaBuilder.buildSize();
        this.pizzaBuilder.buildCheese();
        this.pizzaBuilder.buildPepperoni();
        this.pizzaBuilder.buildMushrooms();
    }
}

// Client Code
const margheritaBuilder = new MargheritaPizzaBuilder();
const director = new PizzaDirector(margheritaBuilder);

director.constructPizza();
const margheritaPizza = margheritaBuilder.getResult();

console.log(margheritaPizza.toString());
